contracts/Market.sol

```diff
@@ -1081,94 +1149,126 @@
+   * @param suppliers The suppliers who own each removal being purchased.
+   */
+  function _transferFunds(
+    bool chargeFee,
     address from,
-    address recipient,
     uint256 countOfRemovalsAllocated,
-    uint256[] memory ids,
-    uint256[] memory amounts,
+    uint256[] memory removalIds,
+    uint256[] memory removalAmounts,
     address[] memory suppliers
   ) internal {
-    uint256[] memory removalIds = ids.slice({
-      from: 0,
-      to: countOfRemovalsAllocated
-    });
-    uint256[] memory removalAmounts = amounts.slice({
-      from: 0,
-      to: countOfRemovalsAllocated
-    });
     uint8 holdbackPercentage;
     uint256 restrictedSupplierFee;
     uint256 unrestrictedSupplierFee;
~    for (uint256 i = 0; i < countOfRemovalsAllocated; ++i) {
~      holdbackPercentage = _removal.getHoldbackPercentage({id: removalIds[i]});
~
~      unrestrictedSupplierFee = removalAmounts[i].mulDiv(_priceMultiple, 100);
~      if (holdbackPercentage > 0) {
~        restrictedSupplierFee = removalAmounts[i].mulDiv(
~          _priceMultiple * holdbackPercentage,
~          10000
~        );
~        unrestrictedSupplierFee -= restrictedSupplierFee;
~        if (
~          _restrictedNORI.getUnderlyingTokenAddress() !=
~          address(_purchasingToken)
~        ) {
~          emit SkipRestrictedNORIERC20Transfer({
~            amount: restrictedSupplierFee,
~            removalId: removalIds[i],
~            currentHoldbackPercentage: holdbackPercentage,
~            rNoriUnderlyingToken: _restrictedNORI.getUnderlyingTokenAddress(),
~            purchasingTokenAddress: address(_purchasingToken)
~          });
~          unrestrictedSupplierFee =
~            unrestrictedSupplierFee +
~            restrictedSupplierFee; // transfer all purchasing token to supplier
~        } else {
~          try
~            _restrictedNORI.mint({
~              amount: restrictedSupplierFee,
~              removalId: removalIds[i]
~            })
~          {
~            // solhint-disable-previous-line no-empty-blocks, Nothing should happen here.
~          } catch {
~            emit RestrictedNORIMintFailure({
~              amount: restrictedSupplierFee,
~              removalId: removalIds[i]
~            });
~          }
~          _purchasingToken.transferFrom({
~            from: from,
~            to: address(_restrictedNORI),
~            amount: restrictedSupplierFee
           });
         }
       }
-      _purchasingToken.transferFrom({
-        from: from,
-        to: _noriFeeWallet,
-        amount: this.calculateNoriFee(removalAmounts[i])
-      });
+      if (chargeFee) {
+        _purchasingToken.transferFrom({
+          from: from,
+          to: _noriFeeWallet,
+          amount: this.calculateNoriFee(removalAmounts[i])
+        });
+      }
       _purchasingToken.transferFrom({
         from: from,
         to: suppliers[i],
         amount: unrestrictedSupplierFee
       });
     }
+  }
```

Comment:
I believe that this is the only legit security issue coming up in the code scanning.  We actually touched on this issue in a meeting but need to follow up with another conversation about how to handle the security issue here.  Basically, the issue is that if one address approves the market to spend its funds, ANY other address could originate a purchase that spends that address's funds, but directs the recipient of the certificate to itself the originator or an address of its choosing.

This is fine with our relayer endpoint because it's permissioned and only intended to be used by an address that we control.  However, with an unpermissioned endpoint, this is an issue that we need to figure out how to address.  It's not a HUGE security concern because the only thing that someone can do with the approved funds is buy carbon removals, which the approval would have been intended for anyway, but it's still an issue for the integrity of our market.

Reply:
My recollection was that we were going to disallow an arbitrary payer on the unpermissioned non-pemit swap.

Reply:
gotcha.  yeah I think I got confused because we talked about potentially having some sort of allowlist permissioning around that but I realize now it would be too involved to implement.  Opened a PR to address this security issue and revert the approval based endpoint to how it was before.  Not sure why I didn't catch a slither warning on the PR that introduced it, but at least we're catching it here.
https://github.com/nori-dot-eco/contracts/pull/500